### ISSUE TITLE
MissingNonWitnessUtxo even before signing

### ISSUE THREAD

**Issue by stevenroose** [bitcoindevkit/bdk#1548]
When you provide `witness_utxo` but not `non_witness_utxo`, previously BDK would complain at sign-time. The solution was in the `SignOptions`:
```rust
			let opts = SignOptions {
				trust_witness_utxo: true,
				..Default::default()
			};
```
But now I am already getting this error at tx creation time when calling `TxBuilder::finish`.

I see that `TxBuilder` has a `only_witness_utxo` option now, which makes this error go away, but which is documented to only affect TxBuilder's internal behavior. I am getting this error when not setting `only_witness_utxo` and by adding foreign inputs.

**Comment by ValuedMammal** [comment]
Would it be possible to come up with a test or code snippet that you believe should work if the bug was resolved?

**Comment by stevenroose** [comment]
Here: https://github.com/stevenroose/bdk/tree/repro-foreign-utxo

So basically, removing that "only_witness_utxo" makes the build step fail. This might be intended behavior? But it's weird that if you don't indicate anything it will complain that non-witness utxo data is missing and you fix it by saying that you will not provide non-witness utxo data?

Before, only the signer would error when there was no non-witness utxo data and you would make the signer not error by adding the `SignOptions` field to make it trust the witness utxo data.

**Comment by stevenroose** [comment]
Like, in the very least, the "only_witness_utxo" method should be re-documented to say "do not require non-witness utxo data" instead of "do not provide non-witness utxo data". Because for foreign utxos, without that method, the builder also only provides witness utxo data but then somehow it is a problem while if you specifically tell you to do the same thing, it is not a problem.

I hope you can see what I'm saying. It might just be ok like this I guess, but it was a regression that wasn't like this before.

**Comment by ValuedMammal** [comment]
As a followup question do you think this logic should be changed to allow both taproot and segwitv0 inputs to _not_ have a `non_witness_utxo`?

https://github.com/bitcoindevkit/bdk/blob/7969898c732a3a387a932a61b1719cedcb6fae3c/crates/wallet/src/wallet/mod.rs#L2170-L2180

**Comment by stevenroose** [comment]
Yeah I think if you already do quite custom stuff like foreign inputs, you should not have too many defaults enforced on you. The signer might refuse to sign it, but that's up to you.

In the `add_foreign_input` method, you have to manually provide a psbt input. So it's a bit stupid to go complain the user didn't provide a certain field in that input, it obviously is doing that on purpose because it's literally manually setting the fields.

It seems that for that method, it wouldn't make sense to error on fields missing in the inputs.
